Ethernet driver improvements: MDIO and PHY drivers

arvinf opened the pull request:
#### Summary

This PR cleans up the Ethernet L2 driver model and encompasses a number of commits that:
1. Add MDIO driver API and a shell
2. Add PHY driver API
3. Add MDIO driver for ATSAM platform
4. Add a generic driver for MII PHYs 
5. Clean-up eth_sam_gmac to use the PHY driver
6. Update `dts` and board files for SoCs in the ATSAM family with Ethernet.

#### Motivation
Presently Ethernet drivers combine MAC functionality, MDIO communication, and PHY state handling all-in-one. This results in code duplication and loss of flexibility for the user. For example user code is not able to communicate with a PHY directly and use additional features like cable fault detection or vendor specific registers, or in case of RGMII setting board specific PHY delays.

This PR adds an MDIO driver, which exposes the MDIO bus of the Ethernet MAC. An MDIO shell and an MDIO driver for the ATSAM platform are also provided. 

The PR also adds a PHY driver API and a generic MII PHY driver. This generic driver is a clean implementation of the PHY handling logic that is currently duplicated across multiple Ethernet drivers. It works with all MII compliant PHYs -- which is the majority of chips on the market. The MII PHY driver uses the SoC specific MDIO bus to access the PHY.  The driver handles PHY polling itself and can be easily extended to work with an interrupt line from the PHY chip. 

The Ethernet MAC driver sets a callback with the PHY driver that is invoked when the link state changes. The PHY driver passes the most recent PHY Link state to that callback, which in turn deals configures the MAC link speed accordingly and signals the higher layers that the link is up or down. 

The connection between the PHY, MAC and MDIO drivers is handled through the board specific `dts` file. e.g.
```
&gmac {
	status = "okay";
	zephyr,random-mac-address;

	phy: phy {
		compatible = "phy,mii";
		status = "okay";
		address = <2>;
		mdio = <&mdio>;
	};
};

&mdio {
	status = "okay";
};
```

The Ethernet driver looks for a child named `phy` which is expected to be a PHY driver; a generic MII PHY driver or a chip specific PHY driver. If the PHY driver needs to communicate over MDIO (as the generic one does) then it is passed a phandle to the MDIO driver node. 

#### Compatibility
Existing Ethernet drivers will continue to work as before. They can be refactored as needed into a separate MDIO driver and the PHY state handling code can be removed. The updated driver can instead call the PHY driver to handle PHY state management.

This PR has been tested on ATSAM platform.

arvinf committed:
boards: atsame54_xpro add support for PHY driver

Update to dts file to use the generic MII PHY driver.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

arvinf committed:
boards: sam4e_xpro: add support for PHY driver

Update to dts file to use the generic MII PHY driver.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

arvinf committed:
boards: sam_e70_xplained: add support for PHY driver

Update to dts file to use the generic MII PHY driver.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

arvinf committed:
soc: arm: atmel: sam4e: add support for MDIO driver

Update to dts file to add MDIO driver.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

arvinf committed:
soc: arm: atmel: same70: add support for MDIO driver

Update to dts file to add MDIO driver.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

arvinf committed:
soc: arm: atmel: same5x: add support for MDIO driver

Update to dts file to add MDIO driver.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

jukkar commented on include/drivers/mdio.h:
```diff
@@ -0,0 +1,150 @@
+ * These are for internal use only, so skip these in
+ * public documentation.
+ */
+__subsystem struct mdio_driver_api {
```
Why `__subsystem` and `__syscalls` below? These should only needed if you want to call these functions from userspace when `CONFIG_USERSPACE` is enabled, not sure what would be the use case for doing that.

jukkar commented on drivers/mdio/CMakeLists.txt:
```diff
@@ -1,3 +1,6 @@
 # SPDX-License-Identifier: Apache-2.0
```
The commit `mdio: Add shell support` contains the actual driver. Please make the shell and the driver to separate commits.

jukkar commented on drivers/ethernet/phy/phy_mii.c:
```diff
@@ -0,0 +1,395 @@
+		return -EIO;
+	}
+
+	bool link_up = bmsr_reg & MII_BMSR_LINK_STATUS;
```
Please introduce variables in the beginning of the block.

jukkar commented on drivers/ethernet/phy/phy_mii.c:
```diff
@@ -0,0 +1,395 @@
+	 * and set link state to up.
+	 */
+	if (cfg->fixed) {
+		static int speed_to_phy_link_speed[] = {
```
Could this be also const?

jukkar commented on include/net/phy.h:
```diff
@@ -0,0 +1,231 @@
+ * These are for internal use only, so skip these in
+ * public documentation.
+ */
+__subsystem struct ethphy_driver_api {
```
Same comment to subsystem and syscalls, these sounds unnecessary unless you have a good use case in mind.

jukkar reviewed (commented):
Nice work, some comments below

arvinf committed:
boards: sam_v71_xult: add support for PHY driver

Update to dts file to use the generic MII PHY driver.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

nandojve commented:
Hello @arvinf ,

Thank you so much for this contribution, nice improvement!

I try run #34718 at on top of b61c23943c and got below:
I used [TagoIO IoT Cloud HTTP Sample](https://docs.zephyrproject.org/latest/samples/net/cloud/tagoio_http_post/README.html).

```
# SAM4E_XPRO
uart:~$ *** Booting Zephyr OS build zephyr-v2.5.0-3023-g74729b7e0c4f  ***
[00:00:00.070,000] <inf> phy_mii: PHY (0) ID 221560
[00:00:00.091,000] <inf> eth_sam: MAC: 02:04:25:a7:7b:80
[00:00:00.091,000] <inf> eth_sam: Queue 0 activated
[00:00:00.097,000] <inf> net_config: Initializing network
[00:00:00.097,000] <inf> net_config: Waiting interface 1 (0x20001508) to be up...
[00:01:33.100,000] <err> net_config: Timeout while waiting network interface
[00:01:33.100,000] <err> net_config: Network initialization failed (-62)
[00:01:33.100,000] <inf> tagoio_http_post: TagoIO IoT - HTTP Client - Temperature demo
[00:01:33.100,000] <inf> tagoio_http_post: Temp: 93
[00:01:38.201,000] <err> tagoio_http_post: Could not resolve dns, error: -11
[00:01:38.201,000] <inf> tagoio_http_post: No connection available.
```

```
# SAM_V71_XULT
[00:00:00.000,000] <inf> i2c_sam_twihs: Device I2C_2 initialized
[00:00:00.000,000] <inf> i2c_sam_twihs: Device I2C_0 initialized
uart:~$ *** Booting Zephyr OS build zephyr-v2.5.0-3023-g74729b7e0c4f  ***
[00:00:00.255,000] <err> mdio_sam: transfer timedout MDIO
[00:00:00.510,000] <err> mdio_sam: transfer timedout MDIO
[00:00:00.765,000] <err> mdio_sam: transfer timedout MDIO
[00:00:00.765,000] <inf> eth_sam: MAC: fc:c2:3d:08:90:35
[00:00:00.765,000] <inf> eth_sam: Queue 0 activated
[00:00:00.765,000] <inf> eth_sam: Queue 1 set to idle
[00:00:00.765,000] <inf> eth_sam: Queue 2 set to idle
[00:00:00.771,000] <inf> net_config: Initializing network
[00:00:00.771,000] <inf> net_config: Waiting interface 1 (0x20401750) to be up...
[00:00:01.520,000] <err> mdio_sam: transfer timedout MDIO
[00:00:02.275,000] <err> mdio_sam: transfer timedout MDIO
...
[00:01:33.642,000] <err> mdio_sam: transfer timedout MDIO
[00:01:33.774,000] <err> net_config: Timeout while waiting network interface
[00:01:33.774,000] <err> net_config: Network initialization failed (-62)
[00:01:33.774,000] <inf> tagoio_http_post: TagoIO IoT - HTTP Client - Temperature demo
[00:01:33.774,000] <inf> tagoio_http_post: Temp: 56
```

Without PR it seems to work normally:

```
# SAM4E_XPRO
[00:00:00.000,000] <inf> eth_sam: MAC: 02:04:25:00:8d:88
[00:00:00.000,000] <inf> eth_sam: Queue 0 activated
[00:00:00.000,000] <inf> eth_sam_phy: Soft Reset of ETH PHY
uart:~$ *** Booting Zephyr OS build zephyr-v2.5.0-2715-g3a4f08cd9047  ***
[00:00:00.090,000] <inf> eth_sam_phy: PHYID: 0x221560 at addr: 0
[00:00:00.096,000] <inf> net_config: Initializing network
[00:00:00.096,000] <inf> net_config: Waiting interface 1 (0x20001508) to be up...
[00:00:02.111,000] <inf> eth_sam: Link up
[00:00:03.804,000] <inf> eth_sam_phy: common abilities: speed 100 Mb, full duplex
[00:00:03.804,000] <inf> net_config: Interface 1 (0x20001508) coming up
[00:00:03.804,000] <inf> net_config: Running dhcpv4 client...
[00:00:06.325,000] <inf> net_dhcpv4: Received: 192.168.2.180
[00:00:06.325,000] <inf> net_config: IPv4 address: 192.168.2.180
[00:00:06.325,000] <inf> net_config: Lease time: 259200 seconds
[00:00:06.325,000] <inf> net_config: Subnet: 255.255.255.0
[00:00:06.325,000] <inf> net_config: Router: 192.168.2.1
[00:00:06.326,000] <inf> tagoio_http_post: TagoIO IoT - HTTP Client - Temperature demo
[00:00:06.326,000] <inf> tagoio_http_post: Temp: 82
[00:00:16.600,000] <inf> tagoio_http_post: Temp: 65
```

```
# SAM_V71_XULT
[00:00:00.000,000] <inf> i2c_sam_twihs: Device I2C_2 initialized
[00:00:00.000,000] <inf> i2c_sam_twihs: Device I2C_0 initialized
uart:~$ *** Booting Zephyr OS build zephyr-v2.5.0-2715-g3a4f08cd9047  ***
[00:00:00.001,000] <inf> eth_sam: MAC: fc:c2:3d:08:90:35
[00:00:00.001,000] <inf> eth_sam: Queue 0 activated
[00:00:00.001,000] <inf> eth_sam: Queue 1 set to idle
[00:00:00.001,000] <inf> eth_sam: Queue 2 set to idle
[00:00:00.001,000] <inf> eth_sam_phy: Soft Reset of ETH PHY
[00:00:00.091,000] <inf> eth_sam_phy: PHYID: 0x221573 at addr: 0
[00:00:00.097,000] <inf> net_config: Initializing network
[00:00:00.097,000] <inf> net_config: Waiting interface 1 (0x20401750) to be up...
[00:00:02.112,000] <inf> eth_sam: Link up
[00:00:03.695,000] <inf> eth_sam_phy: common abilities: speed 100 Mb, full duplex
[00:00:03.695,000] <inf> net_config: Interface 1 (0x20401750) coming up
[00:00:03.695,000] <inf> net_config: Running dhcpv4 client...
[00:00:12.698,000] <inf> net_dhcpv4: Received: 192.168.2.183
[00:00:12.698,000] <inf> net_config: IPv4 address: 192.168.2.183
[00:00:12.698,000] <inf> net_config: Lease time: 259200 seconds
[00:00:12.698,000] <inf> net_config: Subnet: 255.255.255.0
[00:00:12.698,000] <inf> net_config: Router: 192.168.2.1
[00:00:12.698,000] <inf> tagoio_http_post: TagoIO IoT - HTTP Client - Temperature demo
[00:00:12.698,000] <inf> tagoio_http_post: Temp: 87
[00:00:22.976,000] <inf> tagoio_http_post: Temp: 20
[00:00:33.234,000] <inf> tagoio_http_post: Temp: 85
```

How have you been testing?

arvinf committed:
drivers: Add mdio API

This commit adds support for MDIO bus. The bus is used by Ethernet
MACs to communicate with PHYs.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

arvinf committed:
drivers: mdio: add Atmel SAM MDIO driver

MDIO driver for ATSAM platform with Ethernet.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

arvinf committed:
drivers: mdio: add shell

MDIO shell support.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

arvinf committed:
drivers: Add Ethernet PHY API

This commit adds support for Ethernet PHY drivers via a PHY API.

It also includes a driver for a generic MII compliant PHY
which supports most PHYs on the market.

Separating PHY driver from the SoC specific Ethernet driver
simplifies the Ethernet driver code and enables code re-use.
Drivers for specific PHYs with more advanced features, such as
RGMII delay in PHY can be developed independent of the Ethernet
MAC driver.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

arvinf committed:
drivers: eth_sam_gmac: use PHY driver

This commit removes PHY and MDIO specific code from the ATSAM
Ethernet driver and instead relies on a separate PHY driver to
handle the PHY.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

arvinf commented:
@nandojve  I'm testing on a custom board with samE54. I've re-pushed with the a patch that should fix your issue. The PR branch is reconstructed from a larger set of work that is being done and a change was left out. The Ethernet driver on start-up resets the MDIO enable bit which would disable the MDIO driver. This is fixed now. Please try it again with the latest.

nandojve commented:
Hi @arvinf ,

SAM4E is working fine, see below:
```
uart:~$ *** Booting Zephyr OS build zephyr-v2.5.0-3024-g3fe5cced31ed  ***
[00:00:00.070,000] <inf> phy_mii: PHY (0) ID 221560
[00:00:00.091,000] <inf> eth_sam: MAC: 02:04:25:a6:fc:43
[00:00:00.091,000] <inf> eth_sam: Queue 0 activated
[00:00:00.097,000] <inf> net_config: Initializing network
[00:00:00.097,000] <inf> net_config: Waiting interface 1 (0x20001508) to be up...
[00:00:03.310,000] <inf> phy_mii: PHY (0) Link speed 100 Mb, full duplex
[00:00:03.310,000] <inf> eth_sam: Link up
[00:00:03.310,000] <inf> net_config: Interface 1 (0x20001508) coming up
[00:00:03.310,000] <inf> net_config: Running dhcpv4 client...
[00:00:09.831,000] <inf> net_dhcpv4: Received: 192.168.2.178
[00:00:09.832,000] <inf> net_config: IPv4 address: 192.168.2.178
[00:00:09.832,000] <inf> net_config: Lease time: 259200 seconds
[00:00:09.832,000] <inf> net_config: Subnet: 255.255.255.0
[00:00:09.832,000] <inf> net_config: Router: 192.168.2.1
[00:00:09.832,000] <inf> tagoio_http_post: TagoIO IoT - HTTP Client - Temperature demo
[00:00:09.832,000] <inf> tagoio_http_post: Temp: 93
[00:00:20.200,000] <inf> tagoio_http_post: Temp: 28
```

SAM_V71_XULT still with same issue.
```
[00:00:00.000,000] <inf> i2c_sam_twihs: Device I2C_2 initialized
[00:00:00.000,000] <inf> i2c_sam_twihs: Device I2C_0 initialized
uart:~$ *** Booting Zephyr OS build zephyr-v2.5.0-3024-g3fe5cced31ed  ***
[00:00:00.255,000] <err> mdio_sam: transfer timedout MDIO
[00:00:00.510,000] <err> mdio_sam: transfer timedout MDIO
[00:00:00.765,000] <err> mdio_sam: transfer timedout MDIO
[00:00:00.765,000] <inf> eth_sam: MAC: fc:c2:3d:08:90:35
[00:00:00.765,000] <inf> eth_sam: Queue 0 activated
[00:00:00.765,000] <inf> eth_sam: Queue 1 activated
[00:00:00.765,000] <inf> eth_sam: Queue 2 activated
[00:00:00.766,000] <inf> eth_sam: Queue 3 activated
[00:00:00.766,000] <inf> eth_sam: Queue 4 activated
[00:00:00.766,000] <inf> eth_sam: Queue 5 activated
[00:00:00.771,000] <inf> net_config: Initializing network
[00:00:00.771,000] <inf> net_config: Waiting interface 1 (0x20403618) to be up...
[00:00:01.520,000] <err> mdio_sam: transfer timedout MDIO
[00:00:02.275,000] <err> mdio_sam: transfer timedout MDIO
...
[00:01:32.887,000] <err> mdio_sam: transfer timedout MDIO
[00:01:33.642,000] <err> mdio_sam: transfer timedout MDIO
[00:01:33.774,000] <err> net_config: Timeout while waiting network interface
[00:01:33.774,000] <err> net_config: Network initialization failed (-62)
```

nandojve commented:
Hi @arvinf ,

I tried some configs to force phy communication on SAM V7 without success.
I noted that original code at `gmac_init` print default reset value and configure GMAC_NCFGR with proper MCK clock.

```
[00:00:00.000,000] <inf> eth_sam: GMAC_NCFGR at reset: 0x00080000
[00:00:00.000,000] <inf> eth_sam: GMAC_NCFGR: 0x01130040, config: 0x01130040
```

```
diff --git a/drivers/ethernet/eth_sam_gmac.c b/drivers/ethernet/eth_sam_gmac.c
index 71e097d581..1f921f4b4a 100644
--- a/drivers/ethernet/eth_sam_gmac.c
+++ b/drivers/ethernet/eth_sam_gmac.c
@@ -1088,6 +1088,8 @@ static int gmac_init(Gmac *gmac, uint32_t gmac_ncfgr_val)
                return mck_divisor;
        }
 
+       LOG_INF("GMAC_NCFGR at reset: 0x%08x", gmac->GMAC_NCFGR);
+
        /* Set Network Control Register to its default value, clear stats. */
        gmac->GMAC_NCR = GMAC_NCR_CLRSTAT;
 
@@ -1105,6 +1107,8 @@ static int gmac_init(Gmac *gmac, uint32_t gmac_ncfgr_val)
        /* Setup Network Configuration Register */
        gmac->GMAC_NCFGR = gmac_ncfgr_val | mck_divisor;
 
+       LOG_INF("GMAC_NCFGR: 0x%08x, config: 0x%08x", gmac->GMAC_NCFGR, gmac_ncfgr_val | mck_divisor);
+
        gmac->GMAC_UR = DT_ENUM_IDX(DT_DRV_INST(0), phy_connection_type);
 
 #if defined(CONFIG_PTP_CLOCK_SAM_GMAC)
```

and it is made before `phy_sam_gmac_init`
`1972          result = phy_sam_gmac_init(&cfg->phy);`

The MDIO driver start before configure GMAC driver. At that point, no GPIO, MCK etc were configured. I think we could configure that at MDIO driver init, instead at GMAC. I did a try but I don't understand the why NCFGR start as 0x00000000 and don't accept MCK clock divisor.

[00:00:00.000,000] <dbg> mdio_sam.mdio_sam_initialize: MCK: 150000000
[00:00:00.400,000] <dbg> mdio_sam.mdio_sam_initialize: NCFGR: 0x00000000
[00:00:00.400,000] <dbg> mdio_sam.mdio_sam_initialize: NCFGR: 0x00002030
[00:00:00.400,000] <dbg> mdio_sam.mdio_sam_initialize: NCFGR: 0x00002030, mck_div: 0x00100000, val: 0x00102030

@mnkp ,

By any chance you know why NCFGR is init as 0x00000000 and don't accept any clock config?

BTW, I added `soc_pmc_peripheral_enable(ID_GMAC);`

nandojve commented:
Hi @arvinf ,

The problem with SAME7x/SAMV7x happen because gmac register address is wrong at devicetree. The current gmac driver works because it uses header constants from hal but any access to devicetree register return a wrong pointer. The #36749 update current driver version with correct value and replace header constants by devicetree values.

The current PR version still try "talk" with phy before pins and clock are set. This should be reworked to reflect correct flow. I mean, mdio shouldn't have try talk with phy if pins and clocks were not set yet. Maybe move mdio pins and clock initialization to mdio can be the solution.

Rebase on top of #36749 and fix mdio register address for SAME7x/SAMV7x. I would like run a clean version to attest the fix.

arvinf commented:
Thank you for your help. I'll update once #36749 is merged in.

nandojve commented:
> Thank you for your help. I'll update once #36749 is merged in.

Hi @arvinf ,

Just to inform that #36749 was merged.

arvinf commented:
@nandojve - rebased. Could you please test the updated PR with your SAM V7 board?

bdkrae commented:
Hello @arvinf ,

thanks for this great PR! Based on your work I created a solution for the mcux ethernet driver, the WIP can be found [here](https://github.com/bdkrae/zephyr/tree/mcux_mdio_phy_driver). Its working fine, but I experienced some challenges that should be addressed:

- Boot sequence ```MDIO -> PHY -> ETH``` needs to be ensured. So it would be great to have the default priority of PHY set to 70 if possible. Currently it is at a value of 80 which is equal to ETH driver.
- The SOC´s using mcux driver have the mdio communication integrated with the ethernet peripheral registers. Any access to these registers before ETH is fully initialized leads to trouble. I fixed this by adding a KCONFIG-Option named _late_mii_start_. In phy_mii.c it ensures, that phy communication starts when the callback (from ETH-driver) is registered, not at init.
- _MDIO operation finished flag_  is part of the ethernet interrupt and therefore maintained by ETH driver. It could be useful to have a ```transfer_complete```-Function in the MDIO-API.

arvinf commented:
Hi @bdkrae.

Thanks for adding support to the mcux driver.

I'll make the change for the boot priority.

The other changes you requested are specific to handling SoCs where the MDIO registers are shared with Ethernet driver. This isn't always necessarily true and may not be going forward. Conceptually the MDIO is just a bus. Ethernet is a peripheral. They are different things and need different drivers. We'd be making API and design decisions to accomodate architectural decisions made by specific SoC designers to comingle the two into the same register speace. I don't think that is the way to go. It doesn't scale to every sitaution. Also a PHY can use I2C, or SPI, or SMI for communication. This was the reason I created the PR in the first place -- to break the forced coupling of PHY to MDIO. We can't assume that MDIO driver will always be needed or present when Ethernet is.

I think the better approach to accomplish what you want is to have your mcux ethernet driver invoke functions in your mcux MDIO driver directly; for example to complete the MDIO part of an interrupt task. Similarly both MDIO and Ethernet drivers can call the same shared function to initialize the peripheral clocks and registers that is aware of double initalizations. 

It is important to ensure that either driver compiles and works without the other, or at least the Ethernet works without the MDIO driver. So some macro checks or weak functions may be necessary. We want to support cases where the user needs the SoC's ethernet driver but the PHY doesn't use MDIO.

arvinf commented on dts/arm/atmel/same70.dtsi:
```diff
@@ -352,11 +352,18 @@
 		};
 
-		tc0: tc@4000c000 {
+		mdio: mdio@40050088 {
```
The MDIO driver gets the register address from the device tree and it does share the same register space as the Ethernet node. I don't think there is a way to avoid that warning, other than hardcoding the register address into the MDIO driver and putting a throwaway address in the device tree for the MDIO node.

mbolivar-nordic replied:
>  I don't think there is a way to avoid that warning, other than hardcoding the register address into the MDIO driver and putting a throwaway address in the device tree for the MDIO node.

You are free to adjust the dtc warning flags; see e..g https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/nrf5340dk_nrf5340/pre_dt_board.cmake.

mbolivar-nordic replied:
>  if there is no necessity to make a similar tree as below, great!

@nandojve You can definitely make `/soc/foo@deadbeef` and `/soc/bar@deadbeef` nodes with the same unit addresses and skip the warning using pre_dt_board.cmake like I linked to above. Nordic SoCs have the same issue where we have multiple nodes with the same unit address and that's how we silence it.

arvinf replied:
@nandojve - I've kep the node address as `0x40050088` to avoid conflict but kept the `reg` value as `0x40050000` . The `reg` value is picked up by the driver.

mbolivar-nordic replied:
> I've kep the node address as `0x40050088` to avoid conflict but kept the `reg` value as `0x40050000`

Sorry @arvinf, but that's forbidden by the devicetree specification:

> The unit-address must match the first address specified in the reg property of the node.

arvinf replied:
I plan on changing this per suggestion from @nandojve and convert it to a nested structure where the MDIO node is inside the Ethernet node and doesn't need it's own register adderss.

arvinf commented:
@nandojve - Changed MDIO node to be a child of the GMAC instead of a standalone node. It gets the register address from the parent node.

nandojve commented on drivers/mdio/mdio_sam.c:
```diff
@@ -0,0 +1,147 @@
+		if (timeout-- == 0U) {
+			LOG_ERR("transfer timedout %s", DEV_NAME(dev));
+			k_sem_give(&data->sem);
+			return -ETIMEDOUT;
```
1 line space

nandojve commented on drivers/mdio/mdio_sam.c:
```diff
@@ -0,0 +1,147 @@
+		*data_out = cfg->regs->GMAC_MAN & GMAC_MAN_DATA_Msk;
+	}
+
+	k_sem_give(&data->sem);
```
1 line space

nandojve commented on dts/bindings/mdio/mdio-controller.yaml:
```diff
@@ -0,0 +1,22 @@
+    protocol:
+      required: false
+      type: string
+      description:
```
single line comment

nandojve commented on dts/bindings/mdio/mdio-controller.yaml:
```diff
@@ -0,0 +1,22 @@
+        - "clause 22"
+        - "clause 45"
+        - "micrel SMI"
+      default: "clause 22"
```
Every default should be explained at **description**

nandojve commented on include/drivers/mdio.h:
```diff
@@ -0,0 +1,158 @@
+__syscall int mdio_read(const struct device *dev, uint8_t prtad, uint8_t devad,
+			uint16_t *data);
+
+static inline int z_impl_mdio_read(const struct device *dev, uint8_t prtad,
```
Keep function parameters aligned. Review all files
```
static inline int z_impl_mdio_read(const struct device *dev, uint8_t prtad,
				   uint8_t devad, uint16_t *data)
```

nandojve reviewed (changes requested):
Hi @arvinf ,

Thank you for this update. The majority of my comments are alignment styles and I found 2 small issues.

1) I noted that below sequence stop to send data after a `mdio read 0 0` command, **sam4e_xpro** board, see below.
It is only works again after reset.
```
uart:~$ *** Booting Zephyr OS build zephyr-v2.6.0-1825-geb5a803da6f1  ***
[00:00:00.255,000] <err> mdio_sam: transfer timedout MDIO
[00:00:00.510,000] <err> mdio_sam: transfer timedout MDIO
[00:00:00.765,000] <err> mdio_sam: transfer timedout MDIO
[00:00:01.020,000] <err> mdio_sam: transfer timedout MDIO
[00:00:01.020,000] <inf> eth_sam: MAC: 02:04:25:4c:c6:3e
[00:00:01.020,000] <inf> eth_sam: Queue 0 activated
[00:00:01.026,000] <inf> shell_telnet: Telnet shell backend initialized
[00:00:03.734,000] <inf> phy_mii: PHY (0) Link speed 100 Mb, full duplex
[00:00:03.734,000] <inf> eth_sam: Link up
[00:00:03.735,000] <inf> net_telnet_sample: Starting Telnet sample
[00:00:03.735,000] <inf> net_telnet_sample: Running dhcpv4 client...
[00:00:03.735,000] <inf> net_telnet_sample: IPv6 address: 2001:db8::1
[00:00:04.740,000] <inf> net_telnet_sample: IPv4 address: 192.168.2.193
[00:00:04.740,000] <inf> net_telnet_sample: Lease time: 259200 seconds
[00:00:04.740,000] <inf> net_telnet_sample: Subnet: 255.255.255.0
[00:00:04.740,000] <inf> net_telnet_sample: Router: 192.168.2.1
uart:~$ net ping 8.8.8.8
PING 8.8.8.8
28 bytes from 8.8.8.8 to 192.168.2.193: icmp_seq=0 ttl=60 time=16 ms
28 bytes from 8.8.8.8 to 192.168.2.193: icmp_seq=1 ttl=60 time=16 ms
28 bytes from 8.8.8.8 to 192.168.2.193: icmp_seq=2 ttl=60 time=15 ms
uart:~$ mdio read 0 0
0[0]: 0x3000
[00:00:27.978,000] <inf> eth_sam: Link down
uart:~$ net ping 8.8.8.8
PING 8.8.8.8
Cannot send IPv4 ping
uart:~$ 
```

2) When I remove eth cable for 10s on SAMV71 the link not came up and all LEDs stay off.

arvinf commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -1160,14 +1160,15 @@ static int gmac_init(Gmac *gmac, uint32_t gmac_ncfgr_val)
 }
 
-static void link_configure(Gmac *gmac, uint32_t flags)
+static void link_configure(Gmac *gmac, bool full_duplex, bool speed_100M)
```
This is not an API function. Just a local function for this driver.

nandojve commented:
Hi @arvinf ,

It is required a rebase due to conflicts.

nandojve commented on drivers/mdio/mdio_shell.c:
```diff
@@ -0,0 +1,144 @@
+	for (int i = 0; i < 32; i++) {
+		data = 0;
+		if (mdio_read(dev, i, dev_addr, &data) >= 0 &&
+			data != 0xffff) {
```
why not use UINT16_MAX instead 0xffff?
don't forget alignment

nandojve commented on drivers/mdio/mdio_shell.c:
```diff
@@ -0,0 +1,144 @@
+	return 0;
+}
+
+/* mdio write <port_addr> <dev_addr> <data> */
```
not help too much, you can drop it.

arvinf replied:
Do you mean to drop the function or remove the comment?

nandojve replied:
comment, you define those variables and passes to mdio_write

nandojve commented on drivers/mdio/mdio_shell.c:
```diff
@@ -0,0 +1,144 @@
+	return 0;
+}
+
+/* mdio read <port_addr> <dev_addr> */
```
same here

nandojve commented on drivers/mdio/mdio_shell.c:
```diff
@@ -0,0 +1,144 @@
+	if (!dev) {
+		shell_error(sh, "MDIO: Device driver %s not found.",
+			    MDIO_DEVICE);
+		return -ENODEV;
```
1 space for return value, review file.

nandojve commented on drivers/ethernet/phy/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_sources_ifdef(CONFIG_PHY_GENERIC_MII	 phy_mii.c)
```
You need define a library here, see #37512

```
zephyr_library()

zephyr_library_sources_ifdef(CONFIG_PHY_GENERIC_MII	 phy_mii.c)
```

nandojve commented on drivers/ethernet/phy/Kconfig:
```diff
@@ -0,0 +1,47 @@
+	int "Ethernet PHY driver init priority"
+	default 70
+	help
+		Ethernet PHY device driver initialization priority.
```
Fix help text alignment, 2 spaces. Review all

nandojve commented on drivers/ethernet/phy/phy_mii.c:
```diff
@@ -0,0 +1,423 @@
+	uint16_t bmcr_reg = 0;
+	uint16_t bmsr_reg = 0;
+	uint16_t anlpar_reg = 0;
+	uint32_t timeout = CONFIG_PHY_AUTONEG_TIMEOUT_MS / 100;
```
I think it will be better have a DEFINE for this once you use at `k_sleep` same value. Just to avoid misbehavior if in future someone change the code.

nandojve commented on drivers/ethernet/phy/phy_mii.c:
```diff
@@ -0,0 +1,423 @@
+		return -EAGAIN;
+	}
+
+	/* TODO -- invoke link state change callback */
```
Is missing something or it is a future feature?

arvinf replied:
Stale comment

nandojve commented on drivers/ethernet/phy/phy_mii.c:
```diff
@@ -0,0 +1,423 @@
+	.fixed = IS_FIXED_LINK(n),					\
+	.fixed_speed = DT_ENUM_IDX_OR(DT_DRV_INST(n), fixed_link, 0),	\
+	.mdio = UTIL_AND(UTIL_NOT(IS_FIXED_LINK(n)),			\
+			DEVICE_DT_GET(DT_PHANDLE(DT_DRV_INST(n), mdio)))\
```
Alignment

nandojve commented on drivers/ethernet/phy/phy_mii.c:
```diff
@@ -0,0 +1,423 @@
+	PHY_MII_CONFIG(n);						\
+	static struct phy_mii_dev_data phy_mii_dev_data_##n;		\
+	DEVICE_DT_INST_DEFINE(n,					\
+			&phy_mii_initialize,				\
```
Alignment

nandojve commented on dts/bindings/ethernet/phy,mii.yaml:
```diff
@@ -0,0 +1,32 @@
+      required: true
+      description: MDIO driver node
+    no-reset:
+      type: boolean
```
Is this required or not?

nandojve commented on include/net/phy.h:
```diff
@@ -0,0 +1,231 @@
+	/** 100Base-T Half-Duplex */
+	LINK_HALF_100BASE_T		= BIT(2),
+	/** 100Base-T Full-Duplex */
+	LINK_FULL_100BASE_T		= BIT(3),
```
better if you can keep same alignment

arvinf replied:
Not sure what is being asked, specifically.

tbursztyka reviewed (approved):
lgtm

mbolivar-nordic commented on boards/arm/atsame54_xpro/atsame54_xpro.dts:
```diff
@@ -104,4 +104,15 @@ zephyr_udc0: &usb0 {
 	zephyr,random-mac-address;
+
+	phy: phy {
+		compatible = "phy,mii";
```
This is failing CI because `phy` is not a device vendor. The failure is expected. The compatible should be changed.

From the PR description, this is an ethernet PHY, right?

Then please a look at this upstream Linux binding for what to do instead: https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/net/ethernet-phy.yaml

mbolivar-nordic reviewed (commented):
Please fix the binding. I suggest seeing if you can port the Linux ethernet PHY binding to zephyr's bindings language and seeing if you can use a compatible subset of the Linux binding here.

Please recall that devicetree source compatibility with other operating systems is an explicit design goal for Zephyr's use of DTS:

https://docs.zephyrproject.org/latest/guides/dts/design.html#source-compatibility-with-other-operating-systems

mbolivar-nordic reviewed (changes requested):
Meant to hit 'request changes' instead of 'comment'; see above requested changes.

nandojve commented:
> Please fix the binding. I suggest seeing if you can port the Linux ethernet PHY binding to zephyr's bindings language and seeing if you can use a compatible subset of the Linux binding here.
> 
> Please recall that devicetree source compatibility with other operating systems is an explicit design goal for Zephyr's use of DTS:
> 
> https://docs.zephyrproject.org/latest/guides/dts/design.html#source-compatibility-with-other-operating-systems

Hi @mbolivar ,

I think if we follow Linux compatibles about ethernet phy we can get a problem if a board uses two ethernet ports and user wants configure different phy protocols. I'm not sure how differentiate multiple bindings options on a generic driver. Zephyr supports multiple binding on a driver but driver uses Kconfig to differentiate.

I'll suggest to change `phy,mii` to `ethernet-phy` and rename file appropriately.

nandojve reviewed (approved):
LGTM !

decsny commented:
Hi @arvinf , I was just wondering what was the reasoning for putting the phy drivers as a subfolder of the ethernet folder, but mdio as a separate folder?

arvinf commented:
`phy` seemed too generic of a term to keep it directly `drivers`. For example USB 2.0 can have an external phy, and there may be other subsystems in the future with their own phys . In Linux `ethernet`, `mdio`, and `phy` are all kept under `net` which is more clear. Therefore I decided to keep the `phy` directly under `ethernet` since it belongs to that subsystem.

decsny commented:
> `phy` seemed too generic of a term to keep it directly under `drivers`. For example USB 2.0 can have an external phy, and there may be other subsystems in the future with their own phys . In Linux `ethernet`, `mdio`, and `phy` are all kept under `net` which is more clear. Therefore I decided to keep the `phy` directly under `ethernet` since it belongs to that subsystem.

Thanks for your response. I did notice that organization in Linux, too, which is part of why I was wondering about this. I guess it makes sense since zephyr does not have a net folder, so I can understand your point about the phy. I am just curious though, of why not to put mdio also in the ethernet folder?

arvinf commented:
`mdio` is a specific bus protocol, like `spi`. There would be no risk of a conflict as there would be with a generic term like `phy`.

I also considered cases where one would use the `mdio` bus (possibly the bitbang mdio driver) without using an ethernet MAC. For example where a micro would need to control a multiport ethernet switch chip. That's why I left it at top-level.

decsny commented:
> `mdio` is a specific bus protocol, like `spi`. There would be no risk of a conflict as there would be with a generic term like `phy`.
> 
> I also considered cases where one would use the `mdio` bus (possibly the bitbang mdio driver) without using an ethernet MAC. For example where a micro would need to control a multiport ethernet switch chip. That's why I left it at top-level.

Okay, thanks for the info, I understand now
